### PR TITLE
test: let a JUnit extension handle integration test setup DHIS2-16557

### PR DESCRIPTION
**Pull request by teleivo**
This adds a JUnit extension and first uses it in mostly tracker related tests.
https://www.baeldung.com/junit-5-extensions#3-lifecycle-callbacks
https://junit.org/junit5/docs/current/user-guide/#extensions

The JUnit extension replaces the logic of all the different integration test base classes. It does the integration test setup by
* implementing beforeAll, beforeEach, afterEach, afterAll [callbacks](https://www.baeldung.com/junit-5-extensions#3-lifecycle-callbacks)
* detecting if the test is annotated with `@Transactional` and what https://junit.org/junit5/docs/current/user-guide/#writing-tests-test-instance-lifecycle the test has (method [default] or class)

The callbacks run before the corresponding methods in the test instance. This allows us to
* bind the session to the test instance (if needed i.e. no `@Transactional` present)
* setup the admin
* inject it into the security context
* execute the startup routines
* and cleanup afterwards

All of the above will then be done in the correct lifecycle beforeEach/afterEach for tests of lifecycle method and beforeAll/afterAll on tests of lifecycle class.

This means devs can annotate their tests directly with `@Transactional`, 
`@TestInstance(TestInstance.Lifecycle.PER_CLASS)` and set up their data using `@BeforeEach` or `@BeforeAll`.

With this in place we can create a single integration test `@Configuration` that uses this extension. We can then extend it using a single web integration test `@Configuration`.

**Note**: this replicates the logic of the 3 base classes shown below. We will improve the logic later on. We for example might always need to clean the DB even if using `@Transactional` as setup in a `@BeforeAll` is commited https://docs.spring.io/spring-framework/reference/testing/testcontext-framework/tx.html.

## Debugging

Sample debug logs that can be turned on

```
11:53:04.381 [main] DEBUG org.hisp.dhis.test.junit.SpringIntegrationTestExtension - testClass=org.hisp.dhis.icon.IconTest testCallback=beforeAll @Transactional=true testLifecycle=Optional[PER_CLASS] message=ran setUp
11:53:05.119 [main] DEBUG org.hisp.dhis.test.junit.SpringIntegrationTestExtension - testClass=org.hisp.dhis.icon.IconTest testCallback=afterAll @Transactional=true testLifecycle=Optional[PER_CLASS] message=ran tearDown
```

### CASE - transactional: no, lifecycle: method (default)

[IntegrationTestBase](https://github.com/dhis2/dhis2-core/blob/090427ac1e61573b7b1f57d27bc05e11bef1bd82/dhis-2/dhis-test-integration/src/test/java/org/hisp/dhis/test/integration/IntegrationTestBase.java#L49)

BeforeEach:
```java
bindSession();
userService = _userService;
adminUser = preCreateInjectAdminUser();

TestUtils.executeStartupRoutines(applicationContext); // from BaseSpringTest.integrationTestBeforeEach() 
setUpTest(); // from BaseSpringTest.integrationTestBeforeEach() - This is useless as a test class @BeforeEach will be executed after its base class one anyway
```

AfterEach:
```java
// inlined here from BaseSpringTest.nonTransactionalAfter()
clearSecurityContext();
tearDownTest(); // I think we can declare these as @AfterEach in the test itself as it is run before the parents AfterEach https://junit.org/junit5/docs/5.0.1/api/org/junit/jupiter/api/AfterEach.html
try {
  dbmsManager.clearSession();
} catch (Exception e) {
  log.error("Failed to clear hibernate session, reason: {}", e.getMessage(), e);
}
unbindSession();
transactionTemplate.execute(
    status -> {
      dbmsManager.emptyDatabase();
      return null;
});
```

### CASE - transactional: yes, lifecycle: method (default)

[TransactionalIntegrationTest](https://github.com/dhis2/dhis2-core/blob/090427ac1e61573b7b1f57d27bc05e11bef1bd82/dhis-2/dhis-test-integration/src/test/java/org/hisp/dhis/test/integration/TransactionalIntegrationTest.java#L53)

BeforeEach:
```java
userService = _userService;
adminUser = preCreateInjectAdminUser();

TestUtils.executeStartupRoutines(applicationContext); // from BaseSpringTest.integrationTestBeforeEach() 
setUpTest(); // from BaseSpringTest.integrationTestBeforeEach() - This is useless as a test class @BeforeEach will be executed after its base class one anyway
```

AfterEach:
```java
clearSecurityContext();
tearDownTest(); // I think we can declare these as @AfterEach in the test itself as it is run before the parents AfterEach 
try {
  dbmsManager.clearSession();
} catch (Exception e) {
  log.info("Failed to clear hibernate session, reason:" + e.getMessage());
}
```

### CASE - transactional: yes, lifecycle: class

[SingleSetupIntegrationTestBase](https://github.com/dhis2/dhis2-core/blob/090427ac1e61573b7b1f57d27bc05e11bef1bd82/dhis-2/dhis-test-integration/src/test/java/org/hisp/dhis/test/integration/SingleSetupIntegrationTestBase.java#L56)

BeforeAll:
```java
bindSession();

User userAdmin = DhisConvenienceTest.createRandomAdminUserWithEntityManager(entityManager);
adminUsername = userAdmin.getUsername();

XinjectSecurityContextUser(userAdmin);

TestUtils.executeStartupRoutines(applicationContext); // from BaseSpringTest.integrationTestBeforeEach() 
setUpTest(); // from BaseSpringTest.integrationTestBeforeEach() - This is useless as a test class @BeforeEach will be executed after its base class one anyway
```

AfterAll:
```java
// inlined here from BaseSpringTest.nonTransactionalAfter()
clearSecurityContext();
tearDownTest(); // I think we can declare these as @AfterEach in the test itself as it is run before the parents AfterEach https://junit.org/junit5/docs/5.0.1/api/org/junit/jupiter/api/AfterEach.html
try {
  dbmsManager.clearSession();
} catch (Exception e) {
  log.error("Failed to clear hibernate session, reason: {}", e.getMessage(), e);
}
unbindSession();
transactionTemplate.execute(
    status -> {
      dbmsManager.emptyDatabase();
      return null;
});
```

## Next

* roll this out to all integration tests
* can we use a JUnit extension or https://docs.spring.io/spring-framework/reference/testing/testcontext-framework/ctx-management/env-profiles.html#testcontext-ctx-management-env-profiles-ActiveProfilesResolver to solve the blocker https://github.com/spring-projects/spring-framework/issues/26145
* consolidate our test configuration so that web extends the service config

### CODE REVIEW

**Review by enricocolasante — APPROVED** [review]
Looks nice :)

**Review by david-mackessy — APPROVED** [review]

File: dhis-2/dhis-support/dhis-support-test/src/main/java/org/hisp/dhis/test/junit/SpringIntegrationTestExtension.java [hunk 1]
```diff
@@ -0,0 +1,247 @@
...
+ * <ul>
+ *   <li>{@link Lifecycle#PER_METHOD} (default) the above setup and tear down is done using {@link
+ *       BeforeEachCallback} and {@link AfterEachCallback}
+ *   <li>{@link Lifecycle#PER_CLASS} the above setup and tear down is done using {@link
+ *       BeforeAllCallback} and {@link AfterAllCallback}
+ * </ul>
+ */
+@Slf4j
+public class SpringIntegrationTestExtension
+    implements BeforeAllCallback, BeforeEachCallback, AfterAllCallback, AfterEachCallback {
```
Comment (jbee) [inline dhis-2/dhis-support/dhis-support-test/src/main/java/org/hisp/dhis/test/junit/SpringIntegrationTestExtension.java:95]: nice - knowing the tools one is using really helps :)

File: dhis-2/dhis-test-integration/src/test/java/org/hisp/dhis/test/integration/PostgresIntegrationTestBase.java [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+ * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+package org.hisp.dhis.test.integration;
+
+import org.hisp.dhis.test.IntegrationTest;
+import org.hisp.dhis.test.SpringIntegrationTestBase;
+import org.springframework.test.context.ActiveProfiles;
+
+/**
+ * @author Gintare Vilkelyte <vilkelyte.gintare@gmail.com>
```
Comment (jbee) [inline dhis-2/dhis-test-integration/src/test/java/org/hisp/dhis/test/integration/PostgresIntegrationTestBase.java:35]: I think you can add yourself as author ;)

**Review by jbee — APPROVED** [review]
Sweet :100:

**Pull request merged**